webu.c

```diff
@@ -0,0 +1,2197 @@
+ *
+ *
+ *
+ */
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <stddef.h>
+#include <ctype.h>
+#include <locale.h>
+#include "motion.h"
+#include "webu.h"
+#include "translate.h"
+
+
+struct webui_ctx {
+    pthread_mutex_t webu_mutex;  /* The mutex to lock activity on the pipe*/
+
+    int   client_socket;         /* Client socket to the pipe */
+    char *auth;                  /* Authorization provided by user*/
+    char *auth_parms;            /* Authorization parms from config file*/
+
+    char *url;                   /* The URL sent from the client */
+    char *protocol;              /* Protocol provided from the body of request*/
+    char *method;                /* Method provided from the body of the request*/
+
+    char *uri_thread;            /* Parsed thread number from the url*/
+    char *uri_cmd1;              /* Parsed command(action) from the url*/
+    char *uri_cmd2;              /* Parsed command (set) from the url*/
+    char *uri_parm1;             /* Parameter 1 for the command */
+    char *uri_value1;            /* The value for parameter 1*/
+    char *uri_parm2;             /* Parameter 2 for the command */
+    char *uri_value2;            /* The value for parameter 2*/
+
+    char *uri_buffer;            /* Buffer of the header request content */
+
+    char *hostname;              /* Host name provided from header content*/
+    int   cam_count;             /* Count of the number of cameras*/
+    int   cam_threads;           /* Count of the number of camera threads running*/
+    char *lang;                  /* Two character abbreviation for locale language*/
+    char *lang_full;             /* Five character abbreviation for language-country*/
+    char *trans_word[15];        /* The buffer for the translated word or phrase */
+};
+
+
+static void webu_context_init(struct context **cnt, struct webui_ctx *webui) {
+
+    int indx;
+
+    webui->auth_parms    = NULL;
+    webui->method        = NULL;
+    webui->url           = NULL;
+    webui->protocol      = NULL;
+    webui->uri_buffer    = NULL;
+    webui->client_socket = -1;
+    webui->hostname      = NULL;
+
+    /* These will be re-used for multiple calls
+     * so we reserve 512 bytes which should be
+     * plenty (too much?) space and we
+     * don't have into overrun problems
+     */
+
+    webui->method      = mymalloc(10);
+    webui->url         = mymalloc(512);
+    webui->protocol    = mymalloc(10);
+
+    webui->uri_thread  = mymalloc(6);
+    webui->uri_cmd1    = mymalloc(512);
+    webui->uri_cmd2    = mymalloc(512);
+    webui->uri_parm1   = mymalloc(512);
+    webui->uri_value1  = mymalloc(512);
+    webui->uri_parm2   = mymalloc(512);
+    webui->uri_value2  = mymalloc(512);
+
+    webui->uri_buffer = mymalloc(1024);
+    webui->lang       = mymalloc(3);
+    webui->lang_full  = mymalloc(6);
+
+    for (indx=0; indx<=14;indx++){
```

Comment:
Should we use macros instead of raw numbers, like those "14" and "50" (and few lines before "6", "10", "512" etc). From the first sight those numbers do not make any sense, so it is hard to evaluate them against overflow